File: clients/swift/Tests/SynchroTests/ScenarioFixtureLoader.swift
import CryptoKit
import Foundation

enum ScenarioFixtureLoader {
    private struct Catalog: Decodable {
        let schemaVersion: Int
        let scenarios: [Entry]

        enum CodingKeys: String, CodingKey {
            case schemaVersion = "schema_version"
            case scenarios
        }
    }

    private struct Entry: Decodable {
        let scenarioID: String
        let path: String
        let sha256: String

        enum CodingKeys: String, CodingKey {
            case scenarioID = "scenario_id"
            case path
            case sha256
        }
    }

    static func load(id: String) throws -> [String: Any] {
        let root = try repositoryRoot()
        let catalogData = try Data(contentsOf: root.appendingPathComponent("conformance/catalog.json"))
        let catalog = try JSONDecoder().decode(Catalog.self, from: catalogData)
        guard catalog.schemaVersion == 1 else {
            throw loaderError("unsupported scenario catalog schema version")
        }
        let matchingEntries = catalog.scenarios.filter { $0.scenarioID == id }
        guard matchingEntries.count == 1, let entry = matchingEntries.first else {
            throw loaderError("scenario is not present in the catalog: \(id)")
        }

        let resolvedRoot = root.resolvingSymlinksInPath().standardizedFileURL
        let scenarioURL = root.appendingPathComponent(entry.path).resolvingSymlinksInPath().standardizedFileURL
        guard scenarioURL.path.hasPrefix(resolvedRoot.path + "/") else {
            throw loaderError("scenario path escapes the repository: \(id)")
        }
        let data = try Data(contentsOf: scenarioURL)
        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard digest == entry.sha256 else {
            throw loaderError("scenario digest does not match the catalog: \(id)")
        }
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              object["id"] as? String == id else {
            throw loaderError("scenario identity does not match the catalog: \(id)")
        }
        return object
    }

    private static func repositoryRoot() throws -> URL {
        var current = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        for _ in 0..<8 {
            if FileManager.default.fileExists(atPath: current.appendingPathComponent("conformance/catalog.json").path) {
                return current
            }
            current.deleteLastPathComponent()
        }
        throw loaderError("repository root was not found")
    }

    private static func loaderError(_ message: String) -> NSError {
        NSError(domain: "ScenarioFixtureLoader", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    }
}

File: clients/kotlin/synchro/src/test/kotlin/com/trainstar/synchro/ScenarioFixtureLoader.kt
package com.trainstar.synchro

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.security.MessageDigest

internal object ScenarioFixtureLoader {
    @Serializable
    private data class Catalog(
        @kotlinx.serialization.SerialName("schema_version") val schemaVersion: Int,
        val scenarios: List<Entry>,
    )

    @Serializable
    private data class Entry(
        @kotlinx.serialization.SerialName("scenario_id") val scenarioID: String,
        val path: String,
        val sha256: String,
    )

    private val json = Json { ignoreUnknownKeys = false }

    fun load(id: String): JsonObject {
        val root = repositoryRoot()
        val catalog = json.decodeFromString<Catalog>(
            String(Files.readAllBytes(root.resolve("conformance/catalog.json")), Charsets.UTF_8)
        )
        check(catalog.schemaVersion == 1) { "unsupported scenario catalog schema version" }
        val entry = catalog.scenarios.singleOrNull { it.scenarioID == id }
            ?: error("scenario is not present in the catalog: $id")
        val resolvedRoot = root.toRealPath()
        val scenarioPath = root.resolve(entry.path).normalize().toRealPath()
        check(scenarioPath.startsWith(resolvedRoot)) { "scenario path escapes the repository: $id" }
        val data = Files.readAllBytes(scenarioPath)
        val digest = MessageDigest.getInstance("SHA-256").digest(data)
            .joinToString("") { "%02x".format(it.toInt() and 0xff) }
        check(digest == entry.sha256) { "scenario digest does not match the catalog: $id" }
        val scenario = json.parseToJsonElement(data.decodeToString()).jsonObject
        check(scenario["id"]?.jsonPrimitive?.content == id) {
            "scenario identity does not match the catalog: $id"
        }
        return scenario
    }

    private fun repositoryRoot(): Path {
        var current: Path? = Paths.get("").toAbsolutePath().normalize()
        repeat(8) {
            if (Files.exists(current!!.resolve("conformance/catalog.json"))) return current!!
            current = current!!.parent
        }
        error("repository root was not found")
    }
}
